### DISCUSSION TITLE
Important Bug with ControlledEnvironment and Drag&Drop.

### DISCUSSION THREAD

**Discussion by gpisano97** [lukasbach/react-complex-tree#104] [Q&A]
Hello, i'm using your component on React,js (17). I'm working with a big tree (with many nested children), and (with my team mates) we implemented a Controlled Environment (because we need to add and remove items), but the drag and drop logic is broken. We have filled the onDrop event and the "putting an item inside another" mechanism works. If we don't expand any Item also the reordering mechanism works, buf if we expand some Items and we try to move some child became all broken! The "target" element is wrong and in some cases (with the elements on bottom of the tree) also the "item" array is wrong!!! On "bestofreactjs.com" there is the issue #15, and seems to be similar. We are in trouble, this problems is making lose a big amount of time! How can we fix it ? Thanks all for the answers.

Best Regards.

**Comment by madhums** [comment]
> buf if we expand some Items and we try to move some child became all broken! The "target" element is wrong and in some cases (with the elements on bottom of the tree) also the "item" array is wrong!!!

Could you share an example using codesandbox.io to see what you mean by "wrong"? ([for example](https://codesandbox.io/s/react-complex-tree-playground-u5tjs?file=/src/App.tsx))
